R/patch.R

```diff
@@ -0,0 +1,375 @@
+    return(res)
+  } else if (is.data.frame(vec1) && abs_tol != 0) {
+    # (we only need to manually recurse if we potentially have columns that would
+    # be specially processed by the above)
+    if (ncol(vec1) == 0) {
+      rep(TRUE, nrow(vec1))
+    } else {
+      Reduce(`&`, lapply(seq_len(ncol(vec1)), function(col_i) {
+        vec_approx_equal0(vec1[[col_i]], vec2[[col_i]], na_equal, abs_tol, inds1, inds2)
+      }))
+    }
+  } else if (is_bare_list(vec1) && abs_tol != 0) {
+    vapply(seq_along(vec1), function(i) {
+      entry1 <- vec1[[i]]
+      entry2 <- vec2[[i]]
+      vec_size(entry1) == vec_size(entry2) &&
+        # Trying to follow vec_equal: strict on ptypes aside from vec_namedness:
+        identical(
+          vec_set_names(vec_ptype(entry1), NULL),
+          vec_set_names(vec_ptype(entry2), NULL)
+        ) &&
+        all(vec_approx_equal0(entry1, entry2, na_equal, abs_tol))
+    }, logical(1L))
+  } else {
+    # No special handling for any other types/situations. We may want to allow
+    # S3 extension of this method or of a new appropriate vec_proxy_* variant.
+    # See Issue #640.
+    if (!is.null(inds1)) {
+      vec1 <- vec_slice(vec1, inds1)
+    }
+    if (!is.null(inds2)) {
+      vec2 <- vec_slice(vec2, inds2)
+    }
+    # perf: vec1 and vec2 have already been cast to a common ptype; we can't
+    # disable casts, but can say to cast (again...) to that ptype
+    res <- vec_equal(vec1, vec2, na_equal = na_equal, vec_ptype(vec1))
+    return(res)
+  }
+}
+
+#' Calculate compact patch to move from one snapshot/update to another
+#'
+#' @param earlier_snapshot tibble or `NULL`; `NULL` represents that there was no
+#'   data before `later_tbl`.
+#' @param later_tbl tibble; must have the same column names as
+#'   `earlier_snapshot` if it is a tibble.
+#' @param ukey_names character; column names that together, form a unique key
+#'   for `earlier_snapshot` and for `later_tbl`. This is unchecked; see
+#'   [`check_ukey_unique`] if you don't already have this guaranteed.
+#' @param later_format "snapshot" or "update"; default is "snapshot". If
+#'   "snapshot", `later_tbl` will be interpreted as a full snapshot of the data
+#'   set including all ukeys, and any ukeys that are in `earlier_snapshot` but
+#'   not in `later_tbl` are interpreted as deletions, which are currently
+#'   (imprecisely) represented in the output patch as revisions of all
+#'   non-`ukey_names` columns to NA values (using `{vctrs}`). If "update", then
+#'   it's assumed that any deletions have already been represented this way in
+#'   `later_tbl` and any ukeys not in `later_tbl` are simply unchanged; we are
+#'   just ensuring that the update is fully compact for the given
+#'   `compactify_abs_tol`.
+#' @param compactify_abs_tol compactification tolerance; see `apply_compactify`
+#' @return a tibble in compact "update" (diff) format
+#'
+#' @keywords internal
+tbl_diff2 <- function(earlier_snapshot, later_tbl,
```

Comment:
not really important but why isn't this just `tbl_diff`?

Reply:
To attempt to clarify that it's a binary operation, unlike `diff`.

Reply:
right this is a thing that e.g. purrr uses